### PR TITLE
Adding Markers for FUSION HCI Provider Client Platform

### PR DESCRIPTION
**Pull request by suchita-g**

### CODE REVIEW

File: ocs_ci/ocs/constants.py [hunk 1]
```diff
@@ -987,6 +987,8 @@
 RHV_PLATFORM = "rhv"
 ROSA_PLATFORM = "rosa"
 FUSIONAAS_PLATFORM = "fusion_aas"
+HCI_BAREMETAL = "hci_baremetal"
+HCI_VSPHERE = "hci_vsphere"
```
Comment (DanielOsypenko) [inline ocs_ci/ocs/constants.py:991]: just not sure if it is correct.
From OCP perspective the cluster is a BM cluster. And it is not an HCI, I think HCI deployed another way.
image

File: ocs_ci/framework/pytest_customization/marks.py [hunk 2]
```diff
@@ -342,6 +368,27 @@
...
+    config.default_cluster_ctx.ENV_DATA["platform"].lower()
+    in HCI_PROVIDER_CLIENT_PLATFORMS
+    and config.default_cluster_ctx.ENV_DATA["cluster_type"].lower() == "provider",
+    reason="Test will not run on Fusion HCI provider cluster",
+)
+
+skipif_hci_client = pytest.mark.skipif(
+    config.default_cluster_ctx.ENV_DATA["platform"].lower()
+    in HCI_PROVIDER_CLIENT_PLATFORMS
+    and config.default_cluster_ctx.ENV_DATA["cluster_type"].lower() == "hci_client",
```
Comment (ebenahar) [inline ocs_ci/framework/pytest_customization/marks.py:381]: please save "hci_client" as a constant

File: ocs_ci/ocs/cluster.py [hunk 3]
```diff
@@ -2216,6 +2230,20 @@ def is_ms_provider_cluster():
...
+    """
+    Check if the cluster is a Fusion HCI provider cluster
+
+    Returns:
+        bool: True, if the cluster is a Fusion HCI provider cluster. False, otherwise
+
+    """
+    return (
+        config.ENV_DATA["platform"].lower() in constants.HCI_PROVIDER_CLIENT_PLATFORMS
+        and config.ENV_DATA["cluster_type"].lower() == "provider"
```
Comment (ebenahar) [inline ocs_ci/ocs/cluster.py:2243]: please save "provider" as a constant

File: ocs_ci/ocs/constants.py (hunk 1 above)
Comment (suchita-g) [reply ocs_ci/ocs/constants.py:991]: The reason for introducing this type is - when node-specific tests will attempt for execution , we can differentiate the platform API to interact with. Like if we are using vpshere and installing like a baremetal , even though to interact with hardware API we need a vsphere-specific details. Similarly we need credential files for vpshere or baremetal depending on use of the actual base platform

File: ocs_ci/framework/pytest_customization/marks.py [hunk 4]
```diff
@@ -342,6 +368,27 @@
...
+    reason="Test will not run on Fusion HCI client cluster",
+)
+
+skipif_hci_provider_and_client = pytest.mark.skipif(
+    config.ENV_DATA["platform"].lower() in HCI_PROVIDER_CLIENT_PLATFORMS
+    and config.is_hci_provider_exist()
+    and config.is_hci_client_exist(),
```
Comment (ebondare) [inline ocs_ci/framework/pytest_customization/marks.py:388]: So the tests marked with this mark will not be skipped if run on hci provider without clients. Is it the desired behaviour? Wouldn't it be more useful to have a mark for skipping the test on any hci cluster, no matter if it's a provider, client or provider+client multicluster setup?

File: ocs_ci/ocs/constants.py (hunk 1 above)
Comment (suchita-g) [reply ocs_ci/ocs/constants.py:991]: As per my understanding, we are supporting only baremetal platform, at least for phase-1 GA . However, QE are setting cluster using Vsphere or IBMcloud as baremetal cluster . so at the top level it is baremetal platform however H/w interaction is depend on actual platform only

File: ocs_ci/framework/__init__.py [hunk 5]
```diff
@@ -375,6 +375,28 @@ def is_consumer_exist(self):
         cluster_types = [cluster.ENV_DATA["cluster_type"] for cluster in self.clusters]
         return "consumer" in cluster_types
 
+    def is_hci_client_exist(self):
```
Comment (fbalak) [inline ocs_ci/framework/__init__.py:378]: I recommend to rename `is_hci_client_exist` and `is_hci_provider_exist` functions to  `hci_client_exists` and `hci_provider_exists` to improve readability - https://stackoverflow.com/questions/1566745/boolean-method-naming-readability

File: tests/hci_provider_client/test_hci_pc_markers.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+import logging
```
Comment (fbalak) [inline tests/hci_provider_client/test_hci_pc_markers.py:1]: These are all libtest tests. We should move this file or `hci_provider_client` directory to [tests/libtest/](https://github.com/red-hat-storage/ocs-ci/tree/master/tests/libtest).

File: tests/conftest.py [hunk 7]
```diff
@@ -344,7 +344,11 @@ def pytest_collection_modifyitems(session, config, items):
                     log.debug(f"Test: {item} will be skipped due to {skip_condition}")
                     items.remove(item)
                     continue
-            if skipif_upgraded_from_marker:
+            if (
+                skipif_upgraded_from_marker
+                and config.ENV_DATA.get("platform", "").lower()
```
Comment (fbalak) [inline tests/conftest.py]:
```suggestion
                and ocsci_config.ENV_DATA.get("platform", "").lower()
```
This should fix the pr_check error.

**Review by DanielOsypenko — APPROVED** [review]

File: ocs_ci/framework/pytest_customization/marks.py (hunk 4 above)
Comment (suchita-g) [reply ocs_ci/framework/pytest_customization/marks.py]: This is similar to MS function https://github.com/red-hat-storage/ocs-ci/blob/b16f17fbc17c8e774b0fec587fe4a07c3659d191/ocs_ci/framework/pytest_customization/marks.py#L338

this function is also reference from MS Markers

**Review by DanielOsypenko — APPROVED** [review]

File: ocs_ci/framework/pytest_customization/marks.py (hunk 4 above)
Comment (suchita-g) [reply ocs_ci/framework/pytest_customization/marks.py]: Also as now platform type is unique to hci , we can do this in either way

**Review by fbalak — DISMISSED** [review]

**Review by ebenahar — DISMISSED** [review]

**Review by DanielOsypenko — APPROVED** [review]

File: ocs_ci/framework/__init__.py [hunk 8]
```diff
@@ -13,6 +13,7 @@
 import logging
 from collections.abc import Mapping
 from dataclasses import dataclass, field, fields
+from ocs_ci.ocs.constants import HCI_CLIENT, HCI_PROVIDER
```
Comment (suchita-g) [inline ocs_ci/framework/__init__.py:16]: As use of `from ocs_ci.ocs.constants import HCI_CLIENT, HCI_PROVIDER` lib inside `init.py` file causes circular import Error.  with Reference to other code in init.py file , using `hci_client` and `provider` as a hardcoded value.

```ocs_ci/ocs/constants.py:1045: in <module>
[58](https://github.com/red-hat-storage/ocs-ci/actions/runs/6866895930/job/18674124645?pr=8824#step:6:61)
Tue, 14 Nov 2023 19:04:00 GMT
from ocs_ci.utility.utils import get_ocp_version
[59](https://github.com/red-hat-storage/ocs-ci/actions/runs/6866895930/job/18674124645?pr=8824#step:6:62)
Tue, 14 Nov 2023 19:04:00 GMT
ocs_ci/utility/utils.py:41: in <module>
[60](https://github.com/red-hat-storage/ocs-ci/actions/runs/6866895930/job/18674124645?pr=8824#step:6:63)
Tue, 14 Nov 2023 19:04:00 GMT
from ocs_ci.framework import config
[61](https://github.com/red-hat-storage/ocs-ci/actions/runs/6866895930/job/18674124645?pr=8824#step:6:64)
Tue, 14 Nov 2023 19:04:00 GMT
E ImportError: cannot import name 'config' from partially initialized module 'ocs_ci.framework' (most likely due to a circular import) (/home/runner/work/ocs-ci/ocs-ci/ocs_ci/framework/__init__.py)```

**Comment by suchita-g** [comment]
verified: http://magna002.ceph.redhat.com/ocsci-jenkins/openshift-clusters/hcp414-10/hcp414-10_20231113T191750/multicluster/logs/test_report_1700029296.html

**Review by fbalak — APPROVED** [review]

**Review by DanielOsypenko — APPROVED** [review]

**Review by ebenahar — APPROVED** [review]

**Pull request merged**